Keep small bigints small

erights opened the pull request:
This PR only changes the encoding of positive bigints so that positive small bigints have small encodings.

I'm not planning to do the same for small negative bigints, but it would be nice if a similar trick is possible.

FUDCo reviewed (changes requested):
You should update the comment that describes the bigint encoding to reflect this change.

mhofman commented:
After considering alternative encodings, this is a fairly flexible one. It took me a minutes to realize it shortens virtually all non extremely large bigints. I second a comment explaining this.

For future memory. Other encodings are possible using control chars right after the `p` (or `n`) that sort in the right lexical order.

gibson042 commented:
It looks like this is approximating a decimal [Elias delta coding](https://en.wikipedia.org/wiki/Elias_delta_coding), and probably should just go all the way to that. I'll try to provide an example later.

erights commented:
> Elias delta coding

Is it sort-order preserving? The hack they recommend for negatives is not, without mentioning it, so it seems that is not a driving goal.

In any case, our remaining problem is a sort-order preserving encoding on the negative bitints, where small negative bigints have small encodings.

erights commented:
> You should update the comment that describes the bigint encoding to reflect this change.

Done. PTAL.

mhofman commented on packages/store/src/patterns/encodeKey.js:
```diff
@@ -85,17 +85,35 @@ const dbEntryKeyToNumber = k => {
+//   ${length} is the decimal representation of the width (i.e., the count of
+//      digits) of the BigInt.
+//
+//   ${lenlen} is the length of the length, which may only be a single digit.
+//      This limits is to bigints smaller that 10 ** 10 ** 10, which is
+//      beyond what we expect any engine to support anyway. (At some point,
+//      we may well impose an implemention limit much smaller than this.)
```
Makes me realize that we allow longer negative integers than positive ones. `lenlen` max value is `9` where we have a padding of the length to 10 digits for negative numbers. Since `0` is unused for `lenlen`, maybe a quick fix is to encode `${lenlen-1}` ?

mhofman replied:
Also means the negative digit `length` description is wrong, and you can't produce bingints with `11 ** 10` or more.

erights replied:
Done!

erights replied:
> Also means the negative digit length description is wrong, and you can't produce negative bingints with 11 ** 10 or more digits.

Could you suggest a correct description? Thanks.

mhofman commented on packages/store/src/patterns/encodeKey.js:
```diff
@@ -85,17 +85,37 @@ const dbEntryKeyToNumber = k => {
+//   ${lenlenm1} is the length of the length minus 1, which must
+//      only be a single digit. (We subtract 1 because the length of the
+//      length will never be zero.)
+//      This limits us to bigints smaller that 10 ** 10 ** 10, which is
```
I maintain that it's smaller than `11 ** 10 ** 10` ;)

erights replied:
`10 ** 10 ** 11` ;)

Done.

FUDCo reviewed (approved):
It's swell!

gibson042 commented:
> > Elias delta coding
> 
> Is it sort-order preserving? The hack they recommend for negatives is not, without mentioning it, so it seems that is not a driving goal.
> 
> In any case, our remaining problem is a sort-order preserving encoding on the negative bitints, where small negative bigints have small encodings.

Yes, it can be. Here's one possibility that eliminates about three characters from the representation of each positive integer and is symmetrically sized for negative integers: https://jsfiddle.net/0oL65w17/ (run with the console open for visibility)
```js
const bigintToDBEntryKey = n => {
  const nn = n < 0n ? -n : n;
  const nDigits = nn.toString().length;
  const lDigits = nDigits.toString().length;
  if (n < 0n) {
    return `n${

      // A zero for each digit beyond the first
      // in the decimal *count* of decimal digits.
      "0".repeat(lDigits - 1)}${

      // The count of digits, offset to be a
      // number of lDigits digits
      // that never starts with zero.
      // * lDigits=1: nDigits 1-9 => 9-1
      // * lDigits=2: nDigits 10-99 => 99-10
      // * lDigits=3: nDigits 100-999 => 999-100
      // * lDigits=4: nDigits 1000-9999 => 9999-1000
      // * ...
      10**lDigits + 10**(lDigits-1) - 1 - nDigits}${

      // The digits in a complementary representation
      // for reverse sorting.
      (10n**BigInt(nDigits) - 1n + n).toString().padStart(nDigits, "0")
      /*or e.g. 10n**nDigits + 10n**(nDigits-1n) - 1n + n*/}`;
  } else if (n === 0n) {
    return `o`;
  } else {
    return `p${

      // A nine for each digit beyond the first
      // in the decimal *count* of decimal digits.
      "9".repeat(lDigits - 1)}${

      // The count of digits,
      // offset to never start with a nine
      // and padded to lDigits digits.
      // * lDigits=1: nDigits 1-9 => 0-9
      // * lDigits=2: nDigits 10-99 => 00-89
      // * lDigits=3: nDigits 100-999 => 000-899
      // * lDigits=4: nDigits 1000-9999 => 0000-8999
      // * ...
      (nDigits - 10**(lDigits - 1)).toString().padStart(lDigits, "0")}${

      // The digits.
      n}`;
  }
};
```

Control characters could be added for faster human comprehension, but are not necessary.

erights commented:
Hi @gibson042 , https://github.com/Agoric/agoric-sdk/pull/4469#issuecomment-1030770373 is great!

I also appreciate the documentation examples there and the tests at https://jsfiddle.net/o6vw2nfd/ . Would be wonderful to see all this become part of the PR before merging.

I substituted in your code and added TODO placeholders for the rest. Since at this point, the contents of the PR will be almost all your contributions rather than mine, could you take it from here? I added you as co-assignee. Thanks!

erights commented:
Because I'm the original author, github won't let me officially review. But consider me to be added unofficially to the reviewers.

gibson042 commented:
@erights To what extent does compactness matter? These could be made shorter with hexadecimal (and shorter still with base64, needing only four characters total for values up to 4095 and five for values up to 262143), but it would no longer be easy for a human to decode positive values.

erights commented:
Would always be less that 2x difference?

At this stage, I'd say that it need not be shorter. Some human inspectability has some value, though is also imperfect and misleading. At this state, my inclination is to keep the decimal digits and even put back in the second colon that you mention. But not the first, since the numbers to the left of the colon no longer have any obvious relationship to anything.

Any further shrinkage probably should be driven by measurements telling us it would actually make a difference. So not yet.

gibson042 commented on packages/store/src/patterns/encodeKey.js:
```diff
@@ -79,69 +79,101 @@ const dbEntryKeyToNumber = k => {
+// numeric sort order of the corresponding numbers, the characters of the unary
+// prefix are different for negative values (type "n" followed by any number of
+// "#"s [which sort before decimal digits]) vs. positive and zero values (type
+// "p" followed by any number of "~"s [which sort after decimal digits]) and
```
It would also be possible to use a single type character for all bigint values rather than "n" for negative and "p" for nonnegative, e.g.

n | encoded
-- | --
-1n | i#8:8
0n | i~1:0
1n | i~1:1
5n | i~1:5
42n | i~2:42
1234567890n | i~~10:1234567890

gibson042 replied:
There's no significant difference, it just seemed worth raising in case of perceived value from e.g. economy of type prefixes and/or one-to-one mappings. Assuming not, we'll stick with "n" and "p".

erights replied:
good

gibson042 reviewed (commented):
@erights I think this is ready for your review.

erights commented on packages/store/src/patterns/encodeKey.js:
```diff
@@ -45,15 +45,14 @@ const asBits = new BigUint64Array(asNumber.buffer);
     n = 0;
   } else if (is(n, NaN)) {
-    return CanonicalNaN;
+    return `f${CanonicalNaNBits}`;
```
Good choice moving the initial `f` that way.

erights commented on packages/store/test/test-encodeKey.js:
```diff
@@ -55,15 +55,23 @@ const getNaN = (hexEncoding = '0008000000000000') => {
+  [-1234567890n, 'n#89:8765432109'],
+  [-123456789n, 'n0:876543210'],
+  [-1000n, 'n5:8999'],
+  [-999n, 'n6:000'],
```
In my ignorance, I have no opinion. What are the pros and cons?

gibson042 replied:
Basically just subjective ease vs. difficulty of decoding negative numbers as a human. This encoding uses only 90% of all n-digit sequences (because the remaining 10% start with zero and are therefore actually sequences numbers of _fewer_ than n digits), which can therefore be offset according to taste.

n | nines' complement | ten's complement
-- | -- | --
-1n | n8:8 | n9:9
-9n | n8:0 | n9:1
-10n | n7:89 | n8:90
-11n | n7:88 | n8:89
-99n | n7:00 | n8:01
-100n | n6:899 | n7:900
-900n | n6:099 | n7:100
-999n | n6:000 | n7:001
-1234567890n | n#89:8765432109 | n#90:8765432110

erights replied:
10s complement looks easier to me to visually decode. If you agree, let's do that.

gibson042 replied:
Done.

erights reviewed (commented):
LGTM, but I can't approve. However, you already have an approval from @FUDCo . I think it's ok to proceed on that.

Note to @FUDCo , we're using your approval even though substantial changes since then. OTOH, I would approve if I could.

dckc commented on packages/store/test/test-encodeKey.js:
```diff
@@ -115,3 +123,5 @@ test('order invariants', t => {
   }
 });
+
+// TODO Add tests from https://jsfiddle.net/o6vw2nfd/
```
> @dckc @warner @mhofman what do we do for alleged randomness in tests? What should @gibson042 do here?

That looks like property testing. We have 2 precedents for using [fast-check](https://www.npmjs.com/package/fast-check) to generate arbitrary / random inputs.

gibson042 replied:
Thanks, I've added tests after much wrestling with fast-check (and still can't get encodeKey equivalents working, but I'll open a separate PR for that).

gibson042 commented on packages/store/test/test-encodeKey.js:
```diff
@@ -114,3 +124,30 @@ test('order invariants', t => {
+        return assertionPassed(t.true(ea < eb), () => ea < eb);
+      }
+      if (a > b) {
+        return assertionPassed(t.true(ea > eb), () => ea > eb);
```
Reproducing logic like this is super kludgy, but ava assertions returning a success/failure boolean was only added in v4 (cf. https://github.com/avajs/ava/releases/tag/v4.0.0#:~:text=Assertions%20as%20type%20guards ) and the upgrade appears to be non-trivial.

dckc commented on packages/store/test/test-encodeKey.js:
```diff
@@ -114,3 +124,30 @@ test('order invariants', t => {
+test('BigInt encoding comparison corresponds with numeric comparison', async t => {
+  await fc.assert(
+    fc.property(fc.bigInt(), fc.bigInt(), (a, b) => {
+      if (a === b) return assertionPassed(t.pass(), () => true);
+      const ea = encodeKey(a);
+      const eb = encodeKey(b);
+      if (a < b) {
+        return assertionPassed(t.true(ea < eb), () => ea < eb);
+      }
+      if (a > b) {
+        return assertionPassed(t.true(ea > eb), () => ea > eb);
+      }
+      t.fail(`Unexpected incomparable bigints: ${a} ${b}`);
+      return false;
```
is this equivalent and more concise?

```suggestion
      const ea = encodeKey(a);
      const eb = encodeKey(b);
      t.is(a === b, ea === eb);
      t.is(a < b, ea < eb);
      t.is(a > b, ea > eb);
```

gibson042 replied:
Not quite, but I like the general approach. Adopted.

dckc commented on packages/store/test/test-encodeKey.js:
```diff
@@ -114,3 +124,30 @@ test('order invariants', t => {
+  await fc.assert(
+    fc.property(fc.bigInt(), n => {
+      const rt = decodeKey(encodeKey(n));
+      return assertionPassed(t.is(rt, n), () => rt === n);
```
Is it important to return something from the 2nd arg of `fc.property`? Doesn't this suffice?

```suggestion
      t.is(rt, n);
```

gibson042 replied:
The examples at https://github.com/dubzzz/fast-check/#getting-started made it look like returning true(thy) was important, but https://dubzzz.github.io/fast-check/index.html#property suggests that only exactly `false` or an exception are treated as failures. I'll update accordingly.

gibson042 replied:
Hmm, the problem with skipping `return` is that doing so bypasses the fast-check behavior for presenting a reduced counterexample to the property being asserted—rather than something useful like
```
  encodeKey › BigInt encoding comparison corresponds with numeric comparison

  Rejected promise returned by test. Reason:

  Error {
    message: `Property failed after 6 tests␊
    { seed: 1977596688, path: "5:71:124", endOnFailure: true }␊
    Counterexample: [35n,-23836150654801173763396695620220553254412511975658213618617502216148950167540n]␊
    Shrunk 2 time(s)␊
    Got error: Property failed by returning false␊
    ␊
    Hint: Enable verbose mode in order to have the list of all failing values encountered during the run`,
  }
```
we just get
```
  encodeKey › BigInt encoding comparison corresponds with numeric comparison

  Difference:

  - true
  + false
```

I think I'd like to keep `assertionPassed` until it is made superfluous by upgrade to ava v4+ (in which assertions return a "passed" boolean, allowing replacement of `return assertionPassed(t.<method>(…), () => …)` with `return t.<method>(…)`).

dckc replied:
oh wow! Thanks! I was kinda disappointed that fast-check didn't seem to do the usual shrinking thing. Good to know I was just holding it wrong.
